File: sdk/go/client_test.go
package ynlicense

import (
	"context"
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func TestClientActivateAndAPIError(t *testing.T) {
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.URL.Path != "/v1/licenses/activate" || r.Method != http.MethodPost {
			t.Fatalf("unexpected request: %s %s", r.Method, r.URL.Path)
		}
		if got := r.Header.Get("User-Agent"); got != "example-product/2.0" {
			t.Fatalf("unexpected user agent: %q", got)
		}
		var input map[string]any
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
			t.Fatalf("decode request: %v", err)
		}
		if input["app_key"] != "app_test" || input["bind_value"] != "machine-A" {
			t.Fatalf("unexpected request payload: %#v", input)
		}
		w.Header().Set("Content-Type", "application/json")
		if input["card_code"] == "bad" {
			w.WriteHeader(http.StatusNotFound)
			_, _ = w.Write([]byte(`{"success":false,"request_id":"req_error","error":{"code":"CARD_INVALID","message":"card invalid"}}`))
			return
		}
		_, _ = w.Write([]byte(`{"success":true,"request_id":"req_ok","data":{"license_no":"lic_test","status":"active","server_time":"2026-08-10T03:00:00Z"}}`))
	}))
	defer server.Close()

	client, err := NewClient(server.URL+"/", "app_test", WithUserAgent("example-product/2.0"))
	if err != nil {
		t.Fatalf("new client: %v", err)
	}
	result, err := client.Activate(context.Background(), ActivateRequest{
		CardCode:  "YN-TEST",
		BindMode:  "device",
		BindValue: "machine-A",
	})
	if err != nil || result.LicenseNo != "lic_test" || result.ServerTime.IsZero() {
		t.Fatalf("activate: result=%#v err=%v", result, err)
	}

	_, err = client.Activate(context.Background(), ActivateRequest{
		CardCode:  "bad",
		BindMode:  "device",
		BindValue: "machine-A",
	})
	if !IsAPIErrorCode(err, "CARD_INVALID") {
		t.Fatalf("expected CARD_INVALID, got %T %v", err, err)
	}
	apiErr := err.(*APIError)
	if apiErr.HTTPStatus != http.StatusNotFound || apiErr.RequestID != "req_error" {
		t.Fatalf("unexpected API error: %#v", apiErr)
	}
}

func TestClientTimeoutAndConfiguration(t *testing.T) {
	if _, err := NewClient("not-a-url", "app_test"); err == nil {
		t.Fatal("expected invalid base URL error")
	}
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		time.Sleep(100 * time.Millisecond)
		_, _ = w.Write([]byte(`{"success":true,"data":{}}`))
	}))
	defer server.Close()
	client, err := NewClient(server.URL, "app_test", WithHTTPClient(&http.Client{Timeout: 10 * time.Millisecond}))
	if err != nil {
		t.Fatalf("new client: %v", err)
	}
	_, err = client.Heartbeat(context.Background(), HeartbeatRequest{LicenseNo: "lic_test", BindMode: "device", BindValue: "machine-A"})
	if err == nil {
		t.Fatal("expected timeout error")
	}
}

func TestClientLifecycleEndpoints(t *testing.T) {
	seen := make(map[string]int)
	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		seen[r.URL.Path]++
		var input map[string]any
		if err := json.NewDecoder(r.Body).Decode(&input); err != nil {
			t.Fatalf("decode %s request: %v", r.URL.Path, err)
		}
		if input["app_key"] != "app_test" {
			t.Fatalf("%s missing app key: %#v", r.URL.Path, input)
		}
		w.Header().Set("Content-Type", "application/json")
		switch r.URL.Path {
		case "/v1/licenses/verify", "/v1/licenses/renew":
			_, _ = w.Write([]byte(`{"success":true,"data":{"license_no":"lic_test","status":"active","server_time":"2026-08-10T03:00:00Z"}}`))
		case "/v1/licenses/heartbeat":
			_, _ = w.Write([]byte(`{"success":true,"data":{"accepted":true,"server_time":"2026-08-10T03:00:00Z"}}`))
		case "/v1/licenses/unbind":
			_, _ = w.Write([]byte(`{"success":true,"data":{"unbound":true,"license_no":"lic_test","server_time":"2026-08-10T03:00:00Z"}}`))
		default:
			t.Fatalf("unexpected path: %s", r.URL.Path)
		}
	}))
	defer server.Close()
	client, err := NewClient(server.URL, "app_test")
	if err != nil {
		t.Fatalf("new client: %v", err)
	}
	ctx := context.Background()
	if _, err := client.Verify(ctx, VerifyRequest{LicenseNo: "lic_test", BindMode: "device", BindValue: "machine-A"}); err != nil {
		t.Fatalf("verify: %v", err)
	}
	if result, err := client.Heartbeat(ctx, HeartbeatRequest{LicenseNo: "lic_test", BindMode: "device", BindValue: "machine-A"}); err != nil || !result.Accepted {
		t.Fatalf("heartbeat: result=%#v err=%v", result, err)
	}
	if _, err := client.Renew(ctx, RenewRequest{LicenseNo: "lic_test", RenewCardCode: "YN-RENEW", BindMode: "device", BindValue: "machine-A"}); err != nil {
		t.Fatalf("renew: %v", err)
	}
	if result, err := client.Unbind(ctx, UnbindRequest{LicenseNo: "lic_test", BindMode: "device", BindValue: "machine-A"}); err != nil || !result.Unbound {
		t.Fatalf("unbind: result=%#v err=%v", result, err)
	}
	for _, path := range []string{"/v1/licenses/verify", "/v1/licenses/heartbeat", "/v1/licenses/renew", "/v1/licenses/unbind"} {
		if seen[path] != 1 {
			t.Fatalf("expected one request to %s, got %d", path, seen[path])
		}
	}
}

File: sdk/node/test/client.test.js
import assert from "node:assert/strict";
import test from "node:test";

import { LicenseClient, isAPIErrorCode } from "../src/index.js";

test("online client injects app key and preserves API errors", async () => {
  const calls = [];
  const fetch = async (url, options) => {
    const body = JSON.parse(options.body);
    calls.push({ url, options, body });
    if (body.card_code === "bad") {
      return jsonResponse(
        { success: false, request_id: "req_error", error: { code: "CARD_INVALID", message: "card invalid" } },
        { status: 404 },
      );
    }
    return jsonResponse({
      success: true,
      request_id: "req_ok",
      data: { license_no: "lic_test", status: "active", server_time: "2026-08-12T03:00:00Z" },
    });
  };
  const client = new LicenseClient({ baseUrl: "https://license.example.com/", appKey: "app_test", fetch, userAgent: "product/2.0" });
  const result = await client.activate({ card_code: "YN-TEST", bind_mode: "device", bind_value: "machine-A", app_key: "cannot_override" });
  assert.equal(result.license_no, "lic_test");
  assert.equal(calls[0].url, "https://license.example.com/v1/licenses/activate");
  assert.equal(calls[0].body.app_key, "app_test");
  assert.equal(calls[0].options.headers["User-Agent"], "product/2.0");

  await assert.rejects(
    client.activate({ card_code: "bad", bind_mode: "device", bind_value: "machine-A" }),
    (error) => isAPIErrorCode(error, "CARD_INVALID") && error.httpStatus === 404 && error.requestId === "req_error",
  );
});

test("all lifecycle methods use their expected routes", async () => {
  const paths = [];
  const fetch = async (url, options) => {
    const path = new URL(url).pathname;
    const body = JSON.parse(options.body);
    assert.equal(body.app_key, "app_test");
    paths.push(path);
    const data = path.endsWith("heartbeat")
      ? { accepted: true, server_time: "2026-08-12T03:00:00Z" }
      : path.endsWith("unbind")
        ? { unbound: true, license_no: "lic_test", server_time: "2026-08-12T03:00:00Z" }
        : { license_no: "lic_test", status: "active", server_time: "2026-08-12T03:00:00Z" };
    return jsonResponse({ success: true, data });
  };
  const client = new LicenseClient({ baseUrl: "https://license.example.com", appKey: "app_test", fetch });
  const binding = { license_no: "lic_test", bind_mode: "device", bind_value: "machine-A" };
  await client.verify(binding);
  assert.equal((await client.heartbeat(binding)).accepted, true);
  await client.renew({ ...binding, renew_card_code: "YN-RENEW" });
  assert.equal((await client.unbind(binding)).unbound, true);
  assert.deepEqual(paths, [
    "/v1/licenses/verify",
    "/v1/licenses/heartbeat",
    "/v1/licenses/renew",
    "/v1/licenses/unbind",
  ]);
});

test("client enforces timeout, caller cancellation, and response limit", async () => {
  const hangingFetch = (_url, { signal }) => new Promise((_resolve, reject) => {
    if (signal.aborted) reject(signal.reason);
    else signal.addEventListener("abort", () => reject(signal.reason), { once: true });
  });
  const timeoutClient = new LicenseClient({ baseUrl: "https://license.example.com", appKey: "app_test", fetch: hangingFetch, timeoutMs: 10 });
  await assert.rejects(timeoutClient.heartbeat({}), /request failed/i);

  const abortController = new AbortController();
  abortController.abort(new DOMException("cancelled", "AbortError"));
  await assert.rejects(timeoutClient.verify({}, { signal: abortController.signal }), /request failed/i);

  const largeClient = new LicenseClient({
    baseUrl: "https://license.example.com",
    appKey: "app_test",
    fetch: async () => new Response("x", { headers: { "Content-Length": String(2 * 1024 * 1024 + 1) } }),
  });
  await assert.rejects(largeClient.verify({}), /exceeds 2 MiB/);
});

function jsonResponse(body, options = {}) {
  return new Response(JSON.stringify(body), {
    ...options,
    headers: { "Content-Type": "application/json", ...(options.headers || {}) },
  });
}
